src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -319,7 +319,8 @@ private string ConcatAsString<T>(IEnumerable<T> enumerable)
             {
                 cacheFile = FileUtility.SafeRead(_request.Project.RestoreMetadata.CacheFilePath, (stream, path) => CacheFileFormat.Read(stream, _logger, path));
 
-                if (cacheFile.IsValid && StringComparer.Ordinal.Equals(cacheFile.DgSpecHash, newDgSpecHash))
+                if (cacheFile.IsValid && StringComparer.Ordinal.Equals(cacheFile.DgSpecHash, newDgSpecHash) &&
+                    (_request.ExistingLockFile == null || _request.ExistingLockFile.PackageSpec.FilePath.Equals(_request.Project.FilePath)))
```

Comment:
I don't think this is right. 

This does fix your NPE problems....but the lock file not existing also means we cannot no-op.